src/core/net/dnssd_server.cpp

```diff
@@ -1103,8 +1140,39 @@ void Server::FinalizeQuery(QueryTransaction &aQuery, Header::Response aResponseC
     OT_ASSERT(sdType != kDnsQueryNone);
     OT_UNUSED_VARIABLE(sdType);
 
-    mQueryUnsubscribe(mQueryCallbackContext, name);
     aQuery.Finalize(aResponseCode, mSocket);
+    // We are the last subscribing this name
+    if (!HasSameQuery(name, sdType))
+    {
+        mQueryUnsubscribe(mQueryCallbackContext, name);
+    }
+}
+
+bool Server::HasSameQuery(const char *name, DnsQueryType aQueryType)
+{
+    const char *split = GetDomainSplit(name, aQueryType);
+    bool        found = false;
+
+    for (QueryTransaction &query : mQueryTransactions)
+    {
+        char         queryName[Name::kMaxNameSize];
+        DnsQueryType queryType;
+        const char * querySplit;
+
+        if (!query.IsValid())
+        {
+            continue;
+        }
+        queryType  = GetQueryType(query.GetResponseHeader(), query.GetResponseMessage(), queryName);
+        querySplit = GetDomainSplit(queryName, queryType);
```

Comment:
Why do we compare without the domains? 

The external query callbacks do not assume that it's going to be resolved by a Discovery Proxy. So, DNS names with different domains are different in DNS-SD server's viewpoint.